The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21096128609/job/60673806585) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/60673806585) [(plain)](https://github.com/rust-lang/rust/commit/f6c131ba12f5ad5c97a18952714789bb100faf1f/checks/60673806585/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] tests/ui/imports/ambiguous-2.rs ... ok
test [ui] tests/ui/imports/ambiguous-4.rs ... ok
test [ui] tests/ui/imports/ambiguous-7.rs ... ok
test [ui] tests/ui/imports/ambiguous-9.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-glob-vs-multiouter.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-globvsglob.rs ... ok
test [ui] tests/ui/imports/ambiguous-8.rs ... ok
test [ui] tests/ui/imports/ambiguous-glob-vs-expanded-extern.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-no-implicit-prelude.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-core-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-std-glob.rs ... ok
---
---- [ui] tests/ui/async-await/in-trait/bad-signatures.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/async-await/in-trait/bad-signatures/bad-signatures.stderr`
diff of stderr:

- error: expected identifier, found keyword `self`
+ error: expected one of `:`, `@`, or `|`, found keyword `self`
2   --> $DIR/bad-signatures.rs:5:23
3    |
4 LL |     async fn bar(&abc self);

-    |                       ^^^^ expected identifier, found keyword
+    |                       ^^^^ expected one of `:`, `@`, or `|`
6 
- error: expected one of `:`, `@`, or `|`, found keyword `self`
-   --> $DIR/bad-signatures.rs:5:23
+ error[E0185]: method `bar` has a `&self` declaration in the impl, but not in the trait
+   --> $DIR/bad-signatures.rs:11:5
9    |
10 LL |     async fn bar(&abc self);
-    |                  -----^^^^
-    |                  |    |
-    |                  |    expected one of `:`, `@`, or `|`
-    |                  help: declare the type after the parameter binding: `<identifier>: <type>`
+    |     ------------------------ trait method declared without `&self`
+ ...
+ LL |     async fn bar(&self) {}
+    |     ^^^^^^^^^^^^^^^^^^^ `&self` used in impl
15 
16 error: aborting due to 2 previous errors
17 

+ For more information about this error, try `rustc --explain E0185`.
18 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/async-await/in-trait/bad-signatures.rs:11:5
+ error: expected one of `:`, `@`, or `|`, found keyword `self`
+    |                       ^^^^ expected one of `:`, `@`, or `|`
+ error[E0185]: method `bar` has a `&self` declaration in the impl, but not in the trait
+   --> $DIR/bad-signatures.rs:11:5
+    |     ------------------------ trait method declared without `&self`
+ ...
+ LL |     async fn bar(&self) {}
+    |     ^^^^^^^^^^^^^^^^^^^ `&self` used in impl
---
To only update this specific test, also pass `--test-args async-await/in-trait/bad-signatures.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/in-trait/bad-signatures.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/async-await/in-trait/bad-signatures" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021"
stdout: none
--- stderr -------------------------------
error: expected one of `:`, `@`, or `|`, found keyword `self`
##[error]  --> /checkout/tests/ui/async-await/in-trait/bad-signatures.rs:5:23
   |
LL |     async fn bar(&abc self);
   |                       ^^^^ expected one of `:`, `@`, or `|`

error[E0185]: method `bar` has a `&self` declaration in the impl, but not in the trait
##[error]  --> /checkout/tests/ui/async-await/in-trait/bad-signatures.rs:11:5
   |
LL |     async fn bar(&abc self);
   |     ------------------------ trait method declared without `&self`
...
LL |     async fn bar(&self) {}
   |     ^^^^^^^^^^^^^^^^^^^ `&self` used in impl

error: aborting due to 2 previous errors
---
-    |
- LL |     fn foo(&mur Self) {}
-    |                 ^^^^ expected identifier, found keyword
- 
7 error: expected one of `:`, `@`, or `|`, found keyword `Self`
8   --> $DIR/issue-70549-resolve-after-recovered-self-ctor.rs:4:17
9    |

28 LL +     fn bar(&mur Self) {}
29    |
30 
- error: expected identifier, found keyword `Self`
-   --> $DIR/issue-70549-resolve-after-recovered-self-ctor.rs:8:25
-    |
- LL |     fn bar(&'static mur Self) {}
-    |                         ^^^^ expected identifier, found keyword
- 
37 error: expected one of `:`, `@`, or `|`, found keyword `Self`
38   --> $DIR/issue-70549-resolve-after-recovered-self-ctor.rs:8:25
39    |

58 LL +     fn baz(&mut Self @ _) {}
59    |
60 
- error[E0533]: expected unit struct, found self constructor `Self`
-   --> $DIR/issue-70549-resolve-after-recovered-self-ctor.rs:4:17
-    |
- LL |     fn foo(&mur Self) {}
-    |                 ^^^^ not a unit struct
+ error: aborting due to 4 previous errors
66 
- error[E0533]: expected unit struct, found self constructor `Self`
-   --> $DIR/issue-70549-resolve-after-recovered-self-ctor.rs:8:25
-    |
- LL |     fn bar(&'static mur Self) {}
-    |                         ^^^^ not a unit struct
- 
- error: aborting due to 8 previous errors
- 
- For more information about this error, try `rustc --explain E0533`.
76 
---
To only update this specific test, also pass `--test-args parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: expected one of `:`, `@`, or `|`, found keyword `Self`
##[error]  --> /checkout/tests/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs:4:17
   |
LL |     fn foo(&mur Self) {}
   |                 ^^^^ expected one of `:`, `@`, or `|`
   |
help: there is a keyword `mut` with a similar name
   |
LL -     fn foo(&mur Self) {}
LL +     fn foo(&mut Self) {}
   |

error: unexpected lifetime `'static` in pattern
##[error]  --> /checkout/tests/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs:8:13
   |
LL |     fn bar(&'static mur Self) {}
   |             ^^^^^^^
   |
help: remove the lifetime
   |
LL -     fn bar(&'static mur Self) {}
LL +     fn bar(&mur Self) {}
   |

error: expected one of `:`, `@`, or `|`, found keyword `Self`
##[error]  --> /checkout/tests/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs:8:25
   |
LL |     fn bar(&'static mur Self) {}
   |                         ^^^^ expected one of `:`, `@`, or `|`
   |
help: there is a keyword `mut` with a similar name
   |
LL -     fn bar(&'static mur Self) {}
LL +     fn bar(&'static mut Self) {}
   |

error: expected one of `:`, `@`, or `|`, found keyword `Self`
##[error]  --> /checkout/tests/ui/parser/issues/issue-70549-resolve-after-recovered-self-ctor.rs:14:17
   |
LL |     fn baz(&mur Self @ _) {}
   |                 ^^^^ expected one of `:`, `@`, or `|`
   |
help: there is a keyword `mut` with a similar name
   |
LL -     fn baz(&mur Self @ _) {}
LL +     fn baz(&mut Self @ _) {}
   |

error: aborting due to 4 previous errors
------------------------------------------

```

</details>

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -2264,7 +2264,7 @@ impl<'a> Parser<'a> {
             && self.look_ahead(1, |t| *t == token::Comma || *t == token::CloseParen)
         {
             // `fn foo(String s) {}`
-            let ident = self.parse_ident().unwrap();
+            let ident = self.parse_ident_common(true).unwrap();
             let span = pat.span.with_hi(ident.span.hi());
 
             err.span_suggestion(

--- a/compiler/rustc_parse/src/parser/item.rs
+++ b/compiler/rustc_parse/src/parser/item.rs
@@ -408,12 +408,11 @@ impl<'a> Parser<'a> {
         let insert_span = ident_span.shrink_to_lo();
 
         let ident = if self.token.is_ident()
-            && self.token.is_non_reserved_ident()
             && (!is_const || self.look_ahead(1, |t| *t == token::OpenParen))
             && self.look_ahead(1, |t| {
                 matches!(t.kind, token::Lt | token::OpenBrace | token::OpenParen)
             }) {
-            self.parse_ident().unwrap()
+            self.parse_ident_common(true).unwrap()
         } else {
             return Ok(());
         };

--- a/compiler/rustc_parse/src/parser/mod.rs
+++ b/compiler/rustc_parse/src/parser/mod.rs
@@ -469,7 +469,7 @@ impl<'a> Parser<'a> {
         self.parse_ident_common(self.may_recover())
     }
 
-    fn parse_ident_common(&mut self, recover: bool) -> PResult<'a, Ident> {
+    pub(crate) fn parse_ident_common(&mut self, recover: bool) -> PResult<'a, Ident> {
         let (ident, is_raw) = self.ident_or_err(recover)?;
 
         if is_raw == IdentIsRaw::No && ident.is_reserved() {

--- a/tests/ui/parser/macro/kw-in-const-item-pos-recovery-149692.rs
+++ b/tests/ui/parser/macro/kw-in-const-item-pos-recovery-149692.rs
@@ -6,6 +6,7 @@ macro_rules! m {
 }
 
 m!(const Self());
-//~^ ERROR expected one of `!` or `::`, found `(`
+//~^ ERROR expected identifier, found keyword `Self`
+//~^^ ERROR missing `fn` or `struct` for function or struct definition
 
 fn main() {}

--- a/tests/ui/parser/macro/kw-in-const-item-pos-recovery-149692.stderr
+++ b/tests/ui/parser/macro/kw-in-const-item-pos-recovery-149692.stderr
@@ -1,11 +1,22 @@
-error: expected one of `!` or `::`, found `(`
-  --> $DIR/kw-in-const-item-pos-recovery-149692.rs:8:14
+error: expected identifier, found keyword `Self`
+  --> $DIR/kw-in-const-item-pos-recovery-149692.rs:8:10
+   |
+LL | m!(const Self());
+   |          ^^^^ expected identifier, found keyword
+
+error: missing `fn` or `struct` for function or struct definition
+  --> $DIR/kw-in-const-item-pos-recovery-149692.rs:8:10
    |
 LL |     (const $id:item()) => {}
    |            -------- while parsing argument for this `item` macro fragment
 ...
 LL | m!(const Self());
-   |              ^ expected one of `!` or `::`
+   |          ^^^^
+   |
+help: if you meant to call a macro, try
+   |
+LL | m!(const Self!());
+   |              +
 
-error: aborting due to 1 previous error
+error: aborting due to 2 previous errors
 

--- a/tests/ui/parser/macro/kw-in-item-pos-recovery-149692.rs
+++ b/tests/ui/parser/macro/kw-in-item-pos-recovery-149692.rs
@@ -8,12 +8,15 @@ macro_rules! m {
 }
 
 m!(Self());
-//~^ ERROR expected one of `!` or `::`, found `(`
+//~^ ERROR expected identifier, found keyword `Self`
+//~^^ ERROR missing `fn` or `struct` for function or struct definition
 
 m!(Self{});
-//~^ ERROR expected one of `!` or `::`, found `{`
+//~^ ERROR expected identifier, found keyword `Self`
+//~^^ ERROR missing `enum` or `struct` for enum or struct definition
 
 m!(crate());
-//~^ ERROR expected one of `!` or `::`, found `(`
+//~^ ERROR expected identifier, found keyword `crate`
+//~^^ ERROR missing `fn` or `struct` for function or struct definition
 
 fn main() {}

--- a/tests/ui/parser/macro/kw-in-item-pos-recovery-149692.stderr
+++ b/tests/ui/parser/macro/kw-in-item-pos-recovery-149692.stderr
@@ -1,29 +1,57 @@
-error: expected one of `!` or `::`, found `(`
-  --> $DIR/kw-in-item-pos-recovery-149692.rs:10:8
+error: expected identifier, found keyword `Self`
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:10:4
+   |
+LL | m!(Self());
+   |    ^^^^ expected identifier, found keyword
+
+error: missing `fn` or `struct` for function or struct definition
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:10:4
    |
 LL |     ($id:item()) => {}
    |      -------- while parsing argument for this `item` macro fragment
 ...
 LL | m!(Self());
-   |        ^ expected one of `!` or `::`
+   |    ^^^^
+   |
+help: if you meant to call a macro, try
+   |
+LL | m!(Self!());
+   |        +
 
-error: expected one of `!` or `::`, found `{`
-  --> $DIR/kw-in-item-pos-recovery-149692.rs:13:8
+error: expected identifier, found keyword `Self`
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:14:4
+   |
+LL | m!(Self{});
+   |    ^^^^ expected identifier, found keyword
+
+error: missing `enum` or `struct` for enum or struct definition
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:14:4
    |
 LL |     ($id:item()) => {}
    |      -------- while parsing argument for this `item` macro fragment
 ...
 LL | m!(Self{});
-   |        ^ expected one of `!` or `::`
+   |    ^^^^
 
-error: expected one of `!` or `::`, found `(`
-  --> $DIR/kw-in-item-pos-recovery-149692.rs:16:9
+error: expected identifier, found keyword `crate`
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:18:4
+   |
+LL | m!(crate());
+   |    ^^^^^ expected identifier, found keyword
+
+error: missing `fn` or `struct` for function or struct definition
+  --> $DIR/kw-in-item-pos-recovery-149692.rs:18:4
    |
 LL |     ($id:item()) => {}
    |      -------- while parsing argument for this `item` macro fragment
 ...
 LL | m!(crate());
-   |         ^ expected one of `!` or `::`
+   |    ^^^^^
+   |
+help: if you meant to call a macro, try
+   |
+LL | m!(crate!());
+   |         +
 
-error: aborting due to 3 previous errors
+error: aborting due to 6 previous errors
 

--- a/tests/ui/parser/macro/kw-in-item-pos-recovery-151238.rs
+++ b/tests/ui/parser/macro/kw-in-item-pos-recovery-151238.rs
@@ -1,0 +1,13 @@
+//@ edition: 2021
+
+macro_rules! x {
+    ($ty : item) => {};
+}
+x! {
+  trait MyTrait { fn bar(c self) }
+  //~^ ERROR expected identifier, found keyword `self`
+  //~^^ ERROR expected one of `:`, `@`, or `|`, found keyword `self`
+  //~^^^ ERROR expected one of `->`, `;`, `where`, or `{`, found `}`
+}
+
+fn main() {}

--- a/tests/ui/parser/macro/kw-in-item-pos-recovery-151238.stderr
+++ b/tests/ui/parser/macro/kw-in-item-pos-recovery-151238.stderr
@@ -1,0 +1,25 @@
+error: expected identifier, found keyword `self`
+  --> $DIR/kw-in-item-pos-recovery-151238.rs:7:28
+   |
+LL |   trait MyTrait { fn bar(c self) }
+   |                            ^^^^ expected identifier, found keyword
+
+error: expected one of `:`, `@`, or `|`, found keyword `self`
+  --> $DIR/kw-in-item-pos-recovery-151238.rs:7:28
+   |
+LL |   trait MyTrait { fn bar(c self) }
+   |                          --^^^^
+   |                          | |
+   |                          | expected one of `:`, `@`, or `|`
+   |                          help: declare the type after the parameter binding: `<identifier>: <type>`
+
+error: expected one of `->`, `;`, `where`, or `{`, found `}`
+  --> $DIR/kw-in-item-pos-recovery-151238.rs:7:34
+   |
+LL |   trait MyTrait { fn bar(c self) }
+   |                      ---         ^ expected one of `->`, `;`, `where`, or `{`
+   |                      |
+   |                      while parsing this `fn`
+
+error: aborting due to 3 previous errors
+

